Add service: mpdscribble

matthiasbeyer opened the pull request:
This is the first service I'm trying to contribute and I did not test this yet, as I don't know why. So please tell me how to install a service from the nixos package git tree.

Maybe there is also another option how to specify passwords? "goibhniu" on IRC pointed out that I could read the password from a file upon service start, but I don't know how to do this, maybe someone can tell me, too?

matthiasbeyer committed:
Add service: mpdscribble

joachifm commented on nixos/modules/services/audio/mpdscribble.nix:
```diff
@@ -0,0 +1,231 @@
+          description = ''
+            Last.fm Password
+          '';
+        };
```
I believe mpdscribble supports md5sums of passwords. You should encourage users to supply passwords as hashes or provide an option to store authentication info outside the nix store.

matthiasbeyer replied:
How to? And how to test?

joachifm replied:
The simplest is to ask the user to supply the name of the mpdscribble.conf they want to use. By specifying the path as a string, it won't be copied into the store.
You'd run mpdscribble as `mpdscribble -conf cfg.configfile`.

matthiasbeyer replied:
I added a note to the password variable explanation that users should insert MD5 and not plain text passwords.

I guess adding another optional `extraConfig` like file is a bit complicated with the current setup of the service, isn't it?

joachifm replied:
It wouldn't have to be too complicated, but not sure if that level of flexibility is worth it for a service like this. IMO, all options except those related to scrobble auth and which MPD host to use are excessive (I for one don't care about tweaking where the journal is stored or which user/group the service runs as).

matthiasbeyer replied:
> I for one don't care about tweaking where the journal is stored or which user/group the service runs as

Yeah, but others care, I guess... so I included it... if you don't mind! :smile:

joachifm replied:
Oh, sure, I'm just speaking from my own prefence. It's not like my opinion matters, anyway.

joachifm commented on nixos/modules/services/audio/mpdscribble.nix:
```diff
@@ -0,0 +1,231 @@
+        password = mkOption {
+          default = "";
+          example = "";
+          type = type.;
```
Typo here (missing "str")

joachifm commented on nixos/modules/services/audio/mpdscribble.nix:
```diff
@@ -0,0 +1,231 @@
+      script        = "exec mpdscribble";
+      serviceConfig = {
+        User = "mpdscribble";
+        PermissionsStartOnly = true;
```
I think this is redundant; doesn't seem like you're doing any pre-exec setup.

joachifm commented on nixos/modules/services/audio/mpdscribble.nix:
```diff
@@ -0,0 +1,231 @@
+      description   = "MPD scrobble client for lastfm/librefm";
+      wants         = [ "mpd.service" ];
+      path          = [ pkgs.mpdscribble ];
+      script        = "exec mpdscribble";
```
I believe you ought to pass --no-daemon (?) or set serviceConfig.Type = "forking"

matthiasbeyer replied:
Can you tell me for what I need to pass this? I'd like to understand why I need this, you know?

joachifm replied:
mpdscribble forks by default; systemd might think that it has exited successfully, unless you specify Type = "forking". By passing `--no-daemon`, mpdscribble stays in the foreground.

joachifm replied:
Also specifying path and script is a bit excessive here, you might as well just do `serviceConfig.ExecStart = ...`.

matthiasbeyer replied:
I hope I did things right with the latest commit and didn't misunderstand what you said...?

joachifm commented on nixos/modules/services/audio/mpdscribble.nix:
```diff
@@ -0,0 +1,242 @@
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        ExecStart = "${pkgs.mpdscribble}/bin/mpdscribble --no-daemon";
```
It occurred to me that you need to add `--conf ${mpdscribbleConf}` as well, otherwise it will try to load mpdscribble.conf from its homedir.

jagajaga commented:
First of all, squash commits. You are submitting one service, so provide just one commit with it.

jagajaga commented on nixos/modules/services/audio/mpdscribble.nix:
```diff
@@ -0,0 +1,247 @@
+  options = {
+    services.mpdscribble = {
+
+      enable = mkOption {
```
There is mkEnableOption.

jagajaga commented:
Also add your service to nixos/modules/module-list.nix

jagajaga commented on nixos/modules/services/audio/mpdscribble.nix:
```diff
@@ -0,0 +1,247 @@
+with lib;
+
+let
+  uid = config.ids.uids.mpdscribble;
```
Can't see that change in your PR....

jagajaga commented:
And you can test it via editing your configuration.nix and building by 
nixos-rebuild switch -I nixpkgs=/path/to/nixpkgs

matthiasbeyer commented:
> First of all, squash commits. You are submitting one service, so provide just one commit with it.

This is WIP. I'll do my cleanup asap this is _entirely_ ready for submission.

Thanks for the other comments, though. I will fix things tomorrow.

joachifm commented:
FWIW I think this makes more sense as a systemd user service, at least unless you implement something more robust wrt. passwords. MD5 sums are probably too easy to crack to offer any real protection. Even if you can't crack the MD5, you can scrobble on the user's behalf. Of course, if the warning is strong enough, I suppose users can make up their own minds about how important they feel their last.fm account password is (or the ability to scrobble to that account).

lethalman commented:
This service won't start at boot, is that correct?

joachifm commented:
There's no wantedBy and it would have to be started manually, I think.

matthiasbeyer commented:
@lethalman I'm not sure whether it is necessarry. Would be nice to auto-start it if "mpd" gets started (as service). But I don't know how to do it.

I hope I will have some time for this at the weekend.

matthiasbeyer commented:
Hi. I just pushed two small updates on this. I'd like to add more and update to your suggestions, but you need to tell me what to do and (more importantly) _how_ to do it, this is the first service I'm writing and I'm just about to learn it! :smile:

matthiasbeyer committed:
Add mpdscribble in module-list

matthiasbeyer committed:
Add mpdscribble ID

matthiasbeyer commented:
Updated.

Rebased onto latest master and added some more fixes. I don't have any syntax error or something anymore.

Here is my service file for my `configuration.nix`:

```nix
{
  services.mpdscribble = {
    enable = true;
    mpdHost = "localhost";
    mpdPort = 6600;

    lastfm = {
      enable = true;
      username = "<your username>";
      passwort = "<your password md5>";
      journal  = /home/m/.mpdscribble.service.cache;
    };
  };
}
```

Unfortunately, it does not build:

```
building Nix...
building the system configuration...
warning: dumping very large path (> 256 MiB); this may run out of memory
error: file ‘/home/m/.gnupg/S.dirmngr’ has an unsupported type
(use ‘--show-trace’ to show detailed location information)
```

matthiasbeyer commented:
I guess I'll stop working on this. I can't get it working. I'll update my branch with a squash, but then I'll close this PR. starting mpdscribble as user works, so this isn't a major drawback.
